src/common/mem.c

```diff
@@ -1573,4 +1573,428 @@ umem_tx_publish(struct umem_instance *umm, struct umem_rsrvd_act *rsrvd_act)
+	   umem_off_t last_byte)
+{
+	struct umem_cache *cache = store->cache;
+	uint64_t start_bit = (first_byte & UMEM_CACHE_PAGE_SZ_MASK) >> UMEM_CACHE_CHUNK_SZ_SHIFT;
+	uint64_t end_bit   = (last_byte & UMEM_CACHE_PAGE_SZ_MASK) >> UMEM_CACHE_CHUNK_SZ_SHIFT;
+	uint64_t bit_nr;
+	uint64_t bit;
+	uint64_t idx;
+
+	for (bit_nr = start_bit; bit_nr <= end_bit; bit_nr++) {
+		idx = bit_nr >> UMEM_CHUNK_IDX_SHIFT; /** uint64_t index */
+		bit = bit_nr & UMEM_CHUNK_IDX_MASK;
+		page->pg_bmap[idx] |= 1ULL << bit;
+	}
+
+	if (!page->pg_waiting && page->pg_last_checkpoint == page->pg_last_inflight) {
+		/** Keep the page in the waiting list if it's waiting for a transaction to
+		 *  be committed to the WAL before it can be flushed.
+		 */
+		d_list_del(&page->pg_link);
+		d_list_add_tail(&page->pg_link, &cache->ca_pgs_dirty);
+	}
+
+	if (page->pg_last_inflight == wr_tx)
+		return;
+
+	if (store->stor_ops->so_wal_id_cmp(store, wr_tx, page->pg_last_inflight) > 0)
+		page->pg_last_inflight = wr_tx;
+}
+
+int
+umem_cache_touch(struct umem_store *store, uint64_t wr_tx, umem_off_t addr, daos_size_t size)
+{
+	struct umem_cache *cache     = store->cache;
+	struct umem_page  *page;
+	umem_off_t        end_addr  = addr + size - 1;
+	struct umem_page  *end_page;
+	umem_off_t        start_addr;
+
+	if (cache == NULL)
+		return 0; /* TODO: When SMD is supported outside VOS, this will be an error */
+
+	page     = umem_cache_off2page(cache, addr);
+	end_page = umem_cache_off2page(cache, end_addr);
+
+	if (page->pg_copying)
+		return -DER_CHKPT_BUSY;
+
+	if (page != end_page) {
+		/** Eventually, we can just assert equal here.  But until we have a guarantee that
+		 * no allocation will span a page boundary, we have to handle this case.  We should
+		 * never have to span multiple pages though.
+		 */
+		if (end_page->pg_copying)
+			return -DER_CHKPT_BUSY;
+		D_ASSERT((page + 1) == end_page);
+		start_addr = end_addr & ~UMEM_CACHE_PAGE_SZ_MASK;
+
+		touch_page(store, end_page, wr_tx, start_addr, end_addr);
+		end_addr = start_addr - 1;
+	}
+
+	touch_page(store, page, wr_tx, addr, end_addr);
+
+	return 0;
+}
+
+/** We can look into other methods later. For now, only handle a few in-flight pages
+ *  being checkpointed at a time.  This will limit how much memory we use in each
+ *  xstream for checkpointing.
+ */
+#define MAX_INFLIGHT 8
+struct umem_checkpoint_data {
+	d_list_t                 cd_link;
+	struct umem_store_iod    cd_store_iod;
+	d_sg_list_t              cd_sg_list;
+	/** Each page can have at most every other chunk so reserve enough space up front to
+	 *  handle that.
+	 */
+	struct umem_store_region cd_regions[UMEM_CACHE_BMAP_SZ / 2];
+	d_iov_t                  cd_iovs[UMEM_CACHE_BMAP_SZ / 2];
+	/** Handle for the underlying I/O operations */
+	daos_handle_t            cd_fh;
+};
+
+static void
+page2chkpt(struct umem_page *page, struct umem_checkpoint_data *chkpt_data)
```

Comment:
Can we expand this function to generate 'chkpt_data' for multiple pages?

Reply:
It's possible, yes.  SEe the comment above.   One assumption that is simpler with doing this on a single page basis is the size of the umem_chkpoint_data struct is fixed.  If I handle multiple pages, I either have to add some dynamic sizing or split a page or rollback page iod if there isn't enough space.  Anyway, this seemed simpler.

We did discuss this before your break a bit.   One way to handle more than one page at a time is change the post callback to handle more than one fh.

anyway, if you feel it would be better to combine pages, I can certainly look into that.   But this is currently working so I think it may be better to commit to a line in the sand.